Normalize gradients before reduction in DistributedDataParallelC10d

myleott opened the pull request:
Summary: Normalizing by the world size before the reduction is less likely to cause overflow in FP16 training.

Differential Revision: D9594708

apaszke reviewed (changes requested):
Normalization should happen on the coalesced buffers instead of individual parameters

soumith commented:
would be super dope if you added a test for this, so that we dont regress on this in the future.

myleott commented:
@pytorchbot retest this please

teng-li commented:
@myleott agreeing on the above comment, it's super risky to do any DDP change right before our release

myleott commented:
I added the test. But also this is a pretty trivial change and without it fp16 training distributed is much worse.

apaszke reviewed (approved):
This might be important for stability and has a test now, so I'd vote to merge it before the release.

Commit:
Normalize gradients before reduction in DistributedDataParallelC10d (#11109)

Summary:
Normalizing by the world size before the reduction is less likely to cause overflow in FP16 training.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/11109

Differential Revision: D9594708

fbshipit-source-id: 95fe299ce2776d664e6652a05f45d9471a80a326
